wfe2/verify.go

```diff
@@ -18,233 +20,511 @@ import (
+	// defend against the incorrect number of signatures.
+	header := jws.Signatures[0].Header
+	nonce := header.Nonce
+	logEvent.RequestNonce = nonce
+	if len(nonce) == 0 {
+		wfe.stats.Inc("Errors.JWSMissingNonce", 1)
+		logEvent.AddError("JWS is missing an anti-replay nonce")
+		return probs.BadNonce("JWS has no anti-replay nonce")
+	} else if !wfe.nonceService.Valid(nonce) {
+		wfe.stats.Inc("Errors.JWSInvalidNonce", 1)
+		logEvent.AddError("JWS has an invalid anti-replay nonce: %q", nonce)
+		return probs.BadNonce(fmt.Sprintf("JWS has an invalid anti-replay nonce: %q", nonce))
 	}
+	return nil
+}
 
-	// Read body
-	if request.Body == nil {
-		wfe.stats.Inc("Errors.NoPOSTBody", 1)
-		logEvent.AddError("no body on POST")
-		return nil, nil, reg, probs.Malformed("No body on POST")
+// validPOSTURL checks the JWS' URL header against the expected URL based on the
+// HTTP request. This prevents a JWS intended for one endpoint being replayed
+// against a different endpoint. It mutates the provided logEvent to capture any
+// errors.
+func (wfe *WebFrontEndImpl) validPOSTURL(
+	request *http.Request,
+	jws *jose.JSONWebSignature,
+	logEvent *requestEvent) *probs.ProblemDetails {
+	// validPOSTURL is called after parseJWS() which defends against the incorrect
+	// number of signatures.
+	header := jws.Signatures[0].Header
+	extraHeaders := header.ExtraHeaders
+	// Check that there is at least one Extra Header
+	if len(extraHeaders) == 0 {
+		wfe.stats.Inc("Errors.MissingURLinJWS", 1)
+		logEvent.AddError("JWS header parameter 'url' missing")
+		return probs.Malformed("JWS header parameter 'url' required")
+	}
+	// Try to read a 'url' Extra Header as a string
+	headerURL, ok := extraHeaders[jose.HeaderKey("url")].(string)
+	if !ok || len(headerURL) == 0 {
+		wfe.stats.Inc("Errors.MissingURLinJWS", 1)
+		logEvent.AddError("JWS header parameter 'url' missing")
+		return probs.Malformed("JWS header parameter 'url' required")
+	}
+	// Compute the URL we expect to be in the JWS based on the HTTP request
+	expectedURL := url.URL{
+		Scheme: requestProto(request),
+		Host:   request.Host,
+		Path:   request.RequestURI,
+	}
+	// Check that the URL we expect is the one that was found in the signed JWS
+	// header
+	if expectedURL.String() != headerURL {
+		return probs.Malformed(fmt.Sprintf(
+			"JWS header parameter 'url' incorrect. Expected %q got %q",
+			expectedURL.String(), headerURL))
+	}
+	return nil
+}
+
+// parseJWS extracts a JSONWebSignature from an HTTP POST request's body. If
+// there is an error reading the JWS or if it has too few or too many
+// signatures, a problem is returned and the requestEvent is mutated to contain
+// the error.
+func (wfe *WebFrontEndImpl) parseJWS(
+	request *http.Request,
+	logEvent *requestEvent) (*jose.JSONWebSignature, string, *probs.ProblemDetails) {
+	// Verify that the POST request has the expected headers
+	prob := wfe.validPOSTRequest(request, logEvent)
+	if prob != nil {
```

Comment:
Wow I sure forgot about this Go pattern when I wrote this code huh? Thanks :-D fixing across the board.